polkadot/xcm/pallet-xcm/src/precompiles.rs

```diff
@@ -0,0 +1,136 @@
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{Config, VersionedLocation, VersionedXcm, Weight, WeightInfo};
+use alloc::vec::Vec;
+use codec::{DecodeAll, Encode};
+use core::{marker::PhantomData, num::NonZero};
+use pallet_revive::{
+	precompiles::{
+		alloy::{self, sol_types::SolValue},
+		AddressMatcher, Error, Ext, Precompile, RuntimeCosts,
+	},
+	Origin,
+};
+use tracing::error;
+use xcm_executor::traits::WeightBounds;
+
+alloy::sol!("src/precompiles/IXcm.sol");
+use IXcm::*;
+
+pub struct XcmPrecompile<T>(PhantomData<T>);
+
+impl<Runtime> Precompile for XcmPrecompile<Runtime>
+where
+	Runtime: crate::Config + pallet_revive::Config,
+{
+	type T = Runtime;
+	const MATCHER: AddressMatcher = AddressMatcher::Fixed(NonZero::new(10).unwrap());
+	const HAS_CONTRACT_INFO: bool = false;
+	type Interface = IXcm::IXcmCalls;
+
+	fn call(
+		_address: &[u8; 20],
+		input: &Self::Interface,
+		env: &mut impl Ext<T = Self::T>,
+	) -> Result<Vec<u8>, Error> {
+		let origin = env.caller();
+		let frame_origin = match origin {
+			Origin::Root => frame_system::RawOrigin::Root.into(),
+			Origin::Signed(account_id) =>
+				frame_system::RawOrigin::Signed(account_id.clone()).into(),
+		};
+
+		match input {
+			IXcmCalls::xcmSend(IXcm::xcmSendCall { destination, message }) => {
+				let weight = <Runtime as Config>::WeightInfo::send();
+				let _ = env.gas_meter_mut().charge(RuntimeCosts::Precompile(weight))?;
```

Comment:
We added a new function to simplify this.

```suggestion
				let _ = env.charge(weight)?;
```